qiskit/primitives/fidelity/base_fidelity.py

```diff
@@ -0,0 +1,121 @@
+
+        Raises:
+            ValueError: ``left_circuit`` and ``right_circuit`` don't have the same number of qubits.
+        """
+
+        if left_circuit is None or right_circuit is None:
+            self._left_circuit = None
+            self._right_circuit = None
+            self._left_parameters = None
+            self._right_parameters = None
+        else:
+            self.set_circuits(left_circuit, right_circuit)
+
+    @abstractmethod
+    def __call__(
+        self,
+        values_left: np.ndarray | list[np.ndarray] | None = None,
+        values_right: np.ndarray | list[np.ndarray] | None = None,
+    ) -> np.ndarray:
+        """Compute the overlap of two quantum states bound by the
+        parametrizations values_left and values_right.
+
+        Args:
+            values_left: Numerical parameters to be bound to the left circuit.
+            values_right: Numerical parameters to be bound to the right circuit.
+
+        Returns:
+            The overlap of two quantum states defined by two parametrized circuits.
+        """
+        raise NotImplementedError
+
+    def _check_values(
+        self, values: np.ndarray | list[np.ndarray] | None, side: str
+    ) -> np.ndarray | None:
+        """
+        Check whether the passed values match the shape of the parameters of the circuit on the side
+        provided.
+
+        Returns a 2D-Array if values match, `None` if no parameters are passed and raises an error if
+        the shapes don't match.
+        """
+        if side == "left":
+            circuit = self._left_circuit
+        else:
+            circuit = self._right_circuit
+
+        if values is None:
+            if circuit.num_parameters != 0:
+                raise ValueError(
+                    f"`values_{side}` cannot be `None` because the {side} circuit has"
+                    f"{circuit.num_parameters} free parameters."
+                )
+            return None
+        else:
+            return np.atleast_2d(values)
+
+    def set_circuits(self, left_circuit: QuantumCircuit, right_circuit: QuantumCircuit):
+        """
+        Fix the circuits for the fidelity to be computed of.
+        Args:
+            left_circuit: (Parametrized) quantum circuit
+            right_circuit: (Parametrized) quantum circuit
+
+        Raises:
+            ValueError: ``left_circuit`` and ``right_circuit`` don't have the same number of qubits.
+        """
+        if left_circuit.num_qubits != right_circuit.num_qubits:
+            raise ValueError(
+                f"The number of qubits for the left circuit ({left_circuit.num_qubits}) \
+                 and right circuit ({right_circuit.num_qubits}) do not coincide."
+            )
+        # Assigning parameter arrays to the two circuits
+        self._left_parameters = ParameterVector("x", left_circuit.num_parameters)
+        self._left_circuit = left_circuit.assign_parameters(self._left_parameters)
```

Comment:
Maybe this ends up being a no-op in the case the circuit has no parameters. If there a reason not to use the original parameters and create a new circuit with different ones - I think estimator just binds to whats there and does not have equivalent

Reply:
The reason for reassigning parameters is that if `left_circuit == right_circuit`, the parameters are also the same. But we need to assign different numerical values to the left and right circuits, thus, we have to enforce that the parameters differ.
As we have unit tests that pass for the case of circuits without parameters I think it should not be a problem.